File: Arrays/Prefix Sum/3355. Zero Array Transformation I/soln.py
class Solution:
    def isZeroArray(self, nums: List[int], queries: List[List[int]]) -> bool:
        n = len(nums)
        diff = [0]* ( n + 1 )
        for l,r in queries:
            diff[l] += 1
            if(r+1 < len(diff)):
                diff[r+1] -= 1
        availableDecrements = [0] * n
        availableDecrements[0] = diff[0]
        for i in range(1, n):
            availableDecrements[i] = availableDecrements[i-1]+diff[i]
        for i in range(n):
            if(nums[i] > availableDecrements[i]):
                return False
        return True

File: Arrays/Prefix Sum/3355. Zero Array Transformation I/soln.java
class soln {
    public boolean isZeroArray(int[] nums, int[][] queries) {
        int n = nums.length;
        int[] diff = new int[n + 1];
        for (int[] q : queries) {
            int l = q[0], r = q[1];
            diff[l] += 1;
            if (r + 1 < n) {
                diff[r + 1] -= 1;
            }
        }
        int[] availableDecrements = new int[n];
        availableDecrements[0] = diff[0];
        for (int i = 1; i < n; i++) {
            availableDecrements[i] = availableDecrements[i - 1] + diff[i];
        }
        for (int i = 0; i < n; i++) {
            if (nums[i] > availableDecrements[i]) {
                return false;
            }
        }

        return true;
    }
}
